docs/kittengrad.md

```diff
@@ -0,0 +1,151 @@
+
+The read operations (`lookupChain`, `currentBlockchain`) are served locally by
+transactors. The write operations (`insertCanonical`, `updateChain`) initiate
+transactions that are completed asynchronously with confirmation streamed through
+the `journalStream` interface.
+
+When the transactor has built a backlog of transactions, it may
+reject an operation as temporarily unavailable. The client is expected to
+backoff and retry with a randomized exponential backoff or similar mechanism.
+
+The transactor will see through the completion of accepted transactions on a best
+effort basis, which in the absence of transactor failure should result in the eventual
+commit of an accepted transaction.
+The client may wait to confirm a transaction through a streamed block,
+and potentially retry transactions aborted because of transactor failure.
+
+The interface changes from the v0.1 interface are related to the asynchronous nature
+of write operations. Previously, the operations returned a commit receipt in the
+form of a `JournalEntry` immediately.
+Now they return a transaction id, which will be included in some `JournalEntry` in
+a future block streamed through the `journalStream` interface.
+The stream interface itself changes to only emit commited block events instead
+of individual entries.
+
+## Transactor Implementation
+
+### Network Primitives
+
+Transactors form a peer-to-peer network with moderated membership. The network
+manages membership and provides a broadcast primitive used for propagating
+transactions and protocol messages to all transactors.
+
+In a small-scale network, the broadcast primitive can be implemented as direct
+multicast, with each transactor fully connected to all other transactors.
+In a larger partially-connected network, broadcast can be implemented through
+a p2p gossip protocol.
+
+### Transactor State
+
+Each transactor maintains a local copy of the blockchain on disk, and a pointer
+to the last commited block. The transactor also maintains on disk the mediachain
+index which corresponds to the blockchain, mapping canonical to chain references.
+The index is used for creating update transactions and providing read access
+to the mediachain.
+
+In addition to transactions commited in the blockchain, each transactor
+maintains a worklog of pending transactions. The blockchain index together with
+the worklog transactions compute an aggregate index, the working index.
+
+Read requests are serviced by the working index, while write operations are
+initiated on the working index and completed through the transaction protocol.
+
+### Transaction Protocol
+#### Transaction Initiation
+
+In order to perform a write operation, the transactor checks the validity
+of the operation against the working index. Duplicate inserts are rejected,
+as are updates for unknown canonicals. It also checks its worklog backlog,
+against some safety threshold.
+
+The transactor then generates a transaction id and adds the data to the
```

Comment:
Yeah, that's a pretty decent mechanism. I am thinking of deterministically generating insert transaction ids so that we can merge duplicates that have been concurrently accepted. It's mentioned a little further down, perhaps I should specify the update transaction id generation there.